### PR TITLE
Feat: Support When Matched Expression for Merge

### PR DESCRIPTION
**Pull request by eakmanrq**
Users can now override the when matched logic for `INCREMENTAL_BY_UNQIUE_KEY` model kind. Note that macros are not supported in this PR.

### CODE REVIEW

File: sqlmesh/core/dialect.py [hunk 1]
```diff
@@ -294,15 +294,17 @@ def _parse_props(self: Parser) -> t.Optional[exp.Expression]:
     if not key:
         return None
 
-    if self._match(TokenType.L_PAREN):
-        value: t.Optional[exp.Expression] = self.expression(
+    name = key.name.lower()
+    if name == "when_matched":
+        value: t.Optional[exp.Expression] = self._parse_when_matched()[0]
```
Comment (tobymao) [inline sqlmesh/core/dialect.py:299]: can't there be multiple when matched expressions? 

when matched, when not matched etc

File: sqlmesh/core/model/kind.py [hunk 2]
```diff
@@ -208,8 +208,36 @@ def to_expression(self, dialect: str = "", **kwargs: t.Any) -> d.ModelKind:
...
+    @classmethod
+    def _when_matched_validator(cls, v: t.Optional[exp.When]) -> t.Optional[exp.When]:
+        def replace_table_references(expression: exp.Expression) -> exp.Expression:
+            from sqlmesh.core.engine_adapter.base import (
+                MERGE_SOURCE_ALIAS,
+                MERGE_TARGET_ALIAS,
+            )
+
+            if isinstance(expression, exp.Column):
+                if expression.table == "target":
```
Comment (tobymao) [inline sqlmesh/core/model/kind.py:225]: lower

File: sqlmesh/core/model/kind.py [hunk 3]
```diff
@@ -208,8 +208,36 @@ def to_expression(self, dialect: str = "", **kwargs: t.Any) -> d.ModelKind:
...
     unique_key: t.List[exp.Expression]
+    when_matched: t.Optional[exp.When] = None
+
     _unique_key_validator = _unique_key_validator
 
+    @field_validator("when_matched", mode="before")
+    @classmethod
+    def _when_matched_validator(cls, v: t.Optional[exp.When]) -> t.Optional[exp.When]:
+        def replace_table_references(expression: exp.Expression) -> exp.Expression:
+            from sqlmesh.core.engine_adapter.base import (
```
Comment (tobymao) [inline sqlmesh/core/model/kind.py:219]: do we need to set the dialect meta here? look at other places where we validate expressions

File: sqlmesh/core/dialect.py (hunk 1 above)
Comment (eakmanrq) [reply sqlmesh/core/dialect.py:299]: This PR is just implementing when matched. I believe that is all that is needed for now and we can add more later if that is required.

Comment (tobymao) [reply sqlmesh/core/dialect.py:299]: you don't think it's worth it to just implement both? complexity doesn't seem too high

File: sqlmesh/core/model/kind.py (hunk 3 above)
Comment (eakmanrq) [reply sqlmesh/core/model/kind.py:219]: Do you know why the parser doesn't automatically do this when returning expressions?

Comment (eakmanrq) [reply sqlmesh/core/model/kind.py:219]: Also not sure what the purpose of this is and the impact if it is not included.

Comment (eakmanrq) [reply sqlmesh/core/model/kind.py:222]: I did add it.

Comment (tobymao) [reply sqlmesh/core/model/kind.py:222]: parser doesn't add it to every node automatically

**Review by tobymao — APPROVED** [review]

**Pull request merged**